core-processor/src/ext.rs

```diff
@@ -1211,6 +1264,184 @@ mod tests {
+                    MessageContextBuilder::new()
+                        .with_outgoing_limit(u32::MAX)
+                        .build(),
+                )
+                .build(),
+        );
+
+        let data = HandlePacket::default();
+
+        let fake_handle = 0;
+
+        let res = ext.send_push(fake_handle, &[0, 0, 0]);
+        assert_eq!(
+            res.unwrap_err(),
+            FallibleExtError::Core(FallibleExtErrorCore::Message(MessageError::OutOfBounds))
+        );
+
+        let handle = ext.send_init().unwrap();
+
+        let res = ext.send_push(handle, &[1, 2, 3]);
+        assert!(res.is_ok());
+
+        let res = ext.send_push(handle, &[4, 5, 6]);
+        assert!(res.is_ok());
+
+        let large_payload = vec![0u8; MAX_PAYLOAD_SIZE + 1];
+
+        let res = ext.send_push(handle, &large_payload);
+        assert_eq!(
+            res.unwrap_err(),
+            FallibleExtError::Core(FallibleExtErrorCore::Message(
+                MessageError::MaxMessageSizeExceed
+            ))
+        );
+
+        let msg = ext.send_commit(handle, data, 0);
+        assert!(msg.is_ok());
+
+        let res = ext.send_push(handle, &[7, 8, 9]);
+        assert_eq!(
+            res.unwrap_err(),
+            FallibleExtError::Core(FallibleExtErrorCore::Message(MessageError::LateAccess))
+        );
+
+        let (outcome, _) = ext.context.message_context.drain();
+        let ContextOutcomeDrain {
+            mut outgoing_dispatches,
+            ..
+        } = outcome.drain();
+        let dispatch = outgoing_dispatches
+            .pop()
+            .map(|(dispatch, _, _)| dispatch)
+            .expect("Send commit was ok");
+
+        assert_eq!(dispatch.message().payload_bytes(), &[1, 2, 3, 4, 5, 6]);
+    }
+
+    #[test]
+    /// This function tests:
+    ///
+    /// - `send_push_input` on non-existent handle
+    /// - `send_push_input` on valid handle
+    /// - `send_push_input` on used handle
+    /// - `send_push_input` data is added to buffer
+    ///
+    fn test_send_push_input() {
+        let mut ext = Ext::new(
+            ProcessorContextBuilder::new()
+                .with_message_context(
+                    MessageContextBuilder::new()
+                        .with_outgoing_limit(u32::MAX)
+                        .build(),
+                )
+                .build(),
+        );
+
+        let data = HandlePacket::default();
+
+        let fake_handle = 0;
+
+        let res = ext.send_push_input(fake_handle, 0, 1);
+        assert_eq!(
+            res.unwrap_err(),
+            FallibleExtError::Core(FallibleExtErrorCore::Message(MessageError::OutOfBounds))
+        );
+
+        let handle = ext.send_init().unwrap();
+
+        let res = ext
+            .context
+            .message_context
+            .payload_mut()
+            .try_extend_from_slice(&[1, 2, 3, 4, 5, 6]);
+        assert!(res.is_ok());
+
+        let res = ext.send_push_input(handle, 2, 3);
```

Comment:
please add test case when `offset` param of `send_push_input` is greater or equal to `len`.

Reply:
Looking at the inner code, the `len` param of `send_push_input` being greater than `offset` is a possible outcome, and if the `offset` is greater than the length of the data, it just sets the offset and length of the range to 0, therefore, the call to `send_push_input` that I made should not affect the outcome.